### PR TITLE
Sort repo names in PR titles alphabetically

### PR DESCRIPTION
**Pull request by premun**
https://github.com/dotnet/arcade-services/issues/3031

### Release Note Category
- [ ] Feature changes/additions 
- [ ] Bug fixes
- [x] Internal Infrastructure Improvements

### Release Note Description
Fixed a flaky PR title validation in E2E tests

### CODE REVIEW

File: test/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 1]
```diff
@@ -260,19 +260,23 @@ await WaitForUpdatedPullRequestAsync(targetRepoName, targetBranch)
         public async Task CheckBatchedAzDoPullRequest(string source1RepoName, string source2RepoName, string targetRepoName, string targetBranch,
             List<Microsoft.DotNet.DarcLib.DependencyDetail> expectedDependencies, string repoDirectory, bool complete = false)
         {
-            string expectedPRTitle = $"[{targetBranch}] Update dependencies from {_parameters.AzureDevOpsAccount}/{_parameters.AzureDevOpsProject}/{source1RepoName} {_parameters.AzureDevOpsAccount}/{_parameters.AzureDevOpsProject}/{source2RepoName}";
-            await CheckAzDoPullRequest(expectedPRTitle, targetRepoName, targetBranch, expectedDependencies, repoDirectory, complete, true, null, null);
+            string[] expectedPRTitles =
```
Comment (tkapin) [inline test/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:263]: Suggestion: we have just two combinations, but listing all expected PR combinations doesn't capture the intent and wouldn't scale for multiple repos.

What about something like this?
```csharp
pullRequest.Title.starstsWith("[{targetBranch}] Update dependencies from")
foreach (string repoName in [source1RepoName, source2RepoName]) {
    pullRequest.Title.contains($"{_parameters.AzureDevOpsAccount}/{_parameters.AzureDevOpsProject}/{repoName}");
}
```

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 2]
```diff
@@ -686,48 +686,45 @@ private async Task<string> ComputePullRequestTitleAsync(InProgressPullRequest in
...
+            // maxTitleLength = 150 to fit 2/3 repo names in the title
+            const int maxTitleLength = 150;
+            var maxRepoListLength = maxTitleLength - baseTitle.Length;
+            const string delimiter = ", ";
 
-                return titleBuilder.ToString();
-            }
-            else
+            var repoNames = new List<string>();
+            foreach (Guid subscriptionId in uniqueSubscriptionIds)
```
Comment (tkapin) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:711]:
```suggestion
            int titleLength = 0;
            foreach (Guid subscriptionId in uniqueSubscriptionIds)
```

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 3]
```diff
@@ -686,48 +686,45 @@ private async Task<string> ComputePullRequestTitleAsync(InProgressPullRequest in
...
-                return $"[{targetBranch}] Update dependencies to ensure coherency";
+                string repoName = await GetSourceRepositoryAsync(subscriptionId);
+
+                // Strip down repo name.
+                repoName = repoName?.Replace("https://github.com/", null);
+                repoName = repoName?.Replace("https://dev.azure.com/", null);
+                repoName = repoName?.Replace("_git/", null);
+                repoNames.Add(repoName);
+
+                if (repoNames.Sum(n => n.Length + delimiter.Length) > maxRepoListLength)
```
Comment (tkapin) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:721]:
```suggestion
                titleLength += repoName.Length + delimiter.Length;
                if (titleLength > maxRepoListLength)
```

**Review by tkapin — APPROVED** [review]

**Comment by premun** [comment]
Ok, this passed here - https://dev.azure.com/dnceng/internal/_build/results?buildId=2339407&view=results

**Review by MilenaHristova — APPROVED** [review]

File: test/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 4]
```diff
@@ -216,33 +216,33 @@ internal async Task<AsyncDisposableValue<Microsoft.DotNet.DarcLib.PullRequest>>
...
+        public async Task CheckBatchedGitHubPullRequest(string targetBranch, string[] sourceRepoNames,
             string targetRepoName, List<Microsoft.DotNet.DarcLib.DependencyDetail> expectedDependencies, string repoDirectory)
         {
-            string[] expectedPRTitles = [
-                $"[{targetBranch}] Update dependencies from {_parameters.GitHubTestOrg}/{source1RepoName} {_parameters.GitHubTestOrg}/{source2RepoName}",
-                $"[{targetBranch}] Update dependencies from {_parameters.GitHubTestOrg}/{source2RepoName} {_parameters.GitHubTestOrg}/{source1RepoName}",
-            ];
+            var repoNames = sourceRepoNames
+                .Select(name => $"{_parameters.GitHubTestOrg}/{name}")
+                .OrderBy(s => s);
```
Comment (dkurepa) [inline test/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:224]: nit: maybe `r` is better than `s` here, but I don't really mind

Comment (premun) [reply test/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:224]: .NET 8.0 and `Order()` would be the best :)

**Comment by MilenaHristova** [comment]
Rolled out with https://github.com/dotnet/arcade-services/issues/3152

**Pull request merged**